Index: patch/patches/5_url_fix.patch
===================================================================
--- patch/patches/5_url_fix.patch	(revision 0)
+++ patch/patches/5_url_fix.patch	(revision 0)
@@ -0,0 +1,34 @@
+Index: WebKit/Source/WebKit/chromium/src/ChromeClientImpl.cpp
+===================================================================
+--- WebKit/Source/WebKit/chromium/src/ChromeClientImpl.cpp	(revision 78450)
++++ WebKit/Source/WebKit/chromium/src/ChromeClientImpl.cpp	(working copy)
+@@ -257,7 +257,7 @@
+ }
+ 
+ Page* ChromeClientImpl::createWindow(
+-    Frame* frame, const FrameLoadRequest& r, const WindowFeatures& features, const NavigationAction&)
++    Frame* frame, const FrameLoadRequest& r, const WindowFeatures& features, const NavigationAction& action)
+ {
+     if (!m_webView->client())
+         return 0;
+@@ -266,7 +266,7 @@
+     if (!r.resourceRequest().isEmpty())
+         request.bind(r.resourceRequest());
+     WebViewImpl* newView = static_cast<WebViewImpl*>(
+-        m_webView->client()->createView(WebFrameImpl::fromFrame(frame), request, features, r.frameName()));
++        m_webView->client()->createView(WebFrameImpl::fromFrame(frame), action.url(), request, features, r.frameName()));
+     if (!newView)
+         return 0;
+ 
+Index: WebKit/Source/WebKit/chromium/public/WebViewClient.h
+===================================================================
+--- WebKit/Source/WebKit/chromium/public/WebViewClient.h	(revision 78450)
++++ WebKit/Source/WebKit/chromium/public/WebViewClient.h	(working copy)
+@@ -88,6 +88,7 @@
+     // The request parameter is only for the client to check if the request
+     // could be fulfilled.  The client should not load the request.
+     virtual WebView* createView(WebFrame* creator,
++                                const WebURL& url,
+                                 const WebURLRequest& request,
+                                 const WebWindowFeatures& features,
+                                 const WebString& name) {
Index: patch/patch_source.cfg
===================================================================
--- patch/patch_source.cfg	(revision 233)
+++ patch/patch_source.cfg	(working copy)
@@ -3,6 +3,8 @@
 # file entry should be proceeded by the code review or bug report link that it
 # relates to.
 patches = {
+  # http://code.google.com/p/chromiumembedded/issues/detail?id=5
+  '5_url_fix' : '../third_party/',
 }
Index: libcef/browser_webview_delegate.cc
===================================================================
--- libcef/browser_webview_delegate.cc	(revision 233)
+++ libcef/browser_webview_delegate.cc	(working copy)
@@ -164,15 +164,15 @@
 // WebViewClient -------------------------------------------------------------
 
 WebView* BrowserWebViewDelegate::createView(WebFrame* creator,
-    const WebURLRequest& request, const WebWindowFeatures& features,
+    const WebURL& url, const WebURLRequest& request, const WebWindowFeatures& features,
     const WebString& name) {
-  CefString url;
-  if (!request.isNull())
-    url = request.url().spec().utf16();
+  CefString cefUrl;
+  if (!url.isNull())
+    cefUrl = url.spec().utf16();
   CefPopupFeatures cefFeatures;
   TranslatePopupFeatures(features, cefFeatures);
   CefRefPtr<CefBrowserImpl> browser =
-      browser_->UIT_CreatePopupWindow(url, cefFeatures);
+      browser_->UIT_CreatePopupWindow(cefUrl, cefFeatures);
   return browser.get() ? browser->UIT_GetWebView() : NULL;
 }
 
Index: libcef/browser_webview_delegate.h
===================================================================
--- libcef/browser_webview_delegate.h	(revision 233)
+++ libcef/browser_webview_delegate.h	(working copy)
@@ -53,7 +53,7 @@
  public:
   // WebKit::WebViewClient
   virtual WebKit::WebView* createView(
-      WebKit::WebFrame* creator, const WebKit::WebURLRequest& request,
+    WebKit::WebFrame* creator, const WebKit::WebURL& url, const WebKit::WebURLRequest& request,
       const WebKit::WebWindowFeatures& features, const WebKit::WebString& name);
   virtual WebKit::WebWidget* createPopupMenu(WebKit::WebPopupType popup_type);
   virtual WebKit::WebExternalPopupMenu* createExternalPopupMenu(
